# 1
#### (a)

$p_1$ = $\frac{n-1}{n} $

#### (b)

$p_2$ = $\frac{n-1}{n}$

#### (c)

$p_3$ = $(\frac{n-1}{n})^n$

#### (d)

$p_4$ = $1-(\frac{n-1}{n})^n = 1-0.8^5=0.67232$

# 2

In [1]:
from scipy import stats
from statsmodels.formula.api import ols
import pandas as pd 
import numpy as np 

In [2]:
x_data = [1,1,2,2,3,3,4,4,5,5,6]
y_data = [62,67,83,91,122,131,140,144,171,180,170]
df_data = pd.DataFrame({"x":x_data , "y": y_data})
Model_1 = ols("y~x+I(x**2)",data = df_data).fit()
Model_1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.965
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     111.3
Date:                Sun, 13 Dec 2020   Prob (F-statistic):           1.45e-06
Time:                        20:05:06   Log-Likelihood:                -37.832
No. Observations:                  11   AIC:                             81.66
Df Residuals:                       8   BIC:                             82.86
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     21.7636     11.550      1.884      0.096      -4.870      48.397
x             41.3208      7.827      5.280      0.001      23.273      59.369
I(x ** 2)     -2.5065      1.145     -2.189      0.060      -5.147       0.134
==============================================================================
Omnibus:                        0.337   Durbin-Watson:                   2.300
Prob(Omnibus):                  0.845   Jarque-Bera (JB):                0.457
Skew:                           0.233   Prob(JB):                        0.796
Kurtosis:                       2.117   Cond. No.                         90.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## $Model_1: Y = 21.7636+41.3208X-2.5065X^2$

In [3]:
Model_2 = ols("y~I(np.sqrt(x))+I(x**2)",data = df_data).fit()
Model_2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.958
Model:                            OLS   Adj. R-squared:                  0.948
Method:                 Least Squares   F-statistic:                     92.25
Date:                Sun, 13 Dec 2020   Prob (F-statistic):           2.98e-06
Time:                        20:05:06   Log-Likelihood:                -38.826
No. Observations:                  11   AIC:                             83.65
Df Residuals:                       8   BIC:                             84.85
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept       -27.3585     23.077     -1.186      0.270     -80.575      25.858
I(np.sqrt(x))    87.7191     18.727      4.684      0.002      44.536     130.903
I(x ** 2)        -0.1610      0.806     -0.200      0.847      -2.019       1.698
==============================================================================
Omnibus:                        0.108   Durbin-Watson:                   2.191
Prob(Omnibus):                  0.947   Jarque-Bera (JB):                0.325
Skew:                           0.074   Prob(JB):                        0.850
Kurtosis:                       2.171   Cond. No.                         175.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## $Model_2 : Y = -27.3585+87.7191X^{1/2}-0.1610X^2$

conclusion: $model_1$ is better than $model_2$ as the $R^2$ is higher,and the p-value of the coefficient is lower.

In [7]:
import scipy.stats as stats
from sklearn.linear_model import LinearRegression 
from sklearn.model_selection import cross_val_score

x = np.array(x_data).reshape(-1,1)
y = np.array(y_data).reshape(-1,1)
xdf = pd.DataFrame(x,columns=["x"])
xdf["x-square"] = xdf["x"]**2
lr_results1 = cross_val_score(LinearRegression(),xdf,y,cv = 11,scoring = 'neg_mean_squared_error')

print("Model1 MSE:",np.mean(lr_results1))
xdf["x-sqrt"] = np.sqrt(xdf["x"])
xdf = xdf.drop(["x"],axis=1)


lr_results2 = cross_val_score(LinearRegression(),xdf,y,cv = 11,scoring = 'neg_mean_squared_error')
print("Model2 MSE:",np.mean(lr_results2))
print("Choose model1")


Model1 MSE: -167.47243012544055
Model2 MSE: -202.0938873958088
Choose model1


# 3

In [8]:
file = np.loadtxt("/Users/zed/VSCode/Stats-Alg/HW/crab.txt")
df = pd.DataFrame(file)
df = df.rename(columns = {3:"X"})
print("The median value of X :",np.median(df["X"]))

The median value of X : 26.1


In [17]:
# bootstrap confidence intervals
from numpy.random import seed
from numpy.random import rand
from numpy.random import randint
from numpy import percentile
# generate dataset
seed(1)
file = np.loadtxt("/Users/zed/VSCode/Stats-Alg/HW/crab.txt")
df = pd.DataFrame(file)
df = df.rename(columns = {3:"X"})
dataset = list(df[u"X"])
# bootstrap

medians = [0]*1000
scores = []
se = 0
for i in range(1000):
    medians[i]  = np.random.choice(df["X"],size=len(dataset),replace = True)
    scores.append(np.median(medians[i]))
    se += (scores[i]-np.median(dataset))**2

print("SE:",se)

SE: 31.54999999999979


In [15]:
import scipy.stats as stats
mu_es_array=np.array(scores)
mu=np.mean(mu_es_array)
std=stats.sem(mu_es_array)
print(stats.norm.interval(0.95, loc=mu, scale=std))

(26.09689639333184, 26.118903606668162)


# 4

Error $= \sum_{i=1}^n y_ilog(h(x_i))+(1-y_i)log(1-h(x_i))$


$h(x)=\frac{exp(\beta_0+\beta_1x)}{1+exp(\beta_0+\beta_1x)}$